lib/engine/game/g_18_norway/game.rb

```diff
@@ -121,26 +121,31 @@ def mountain?(hex)
         end
 
         def route_cost(route)
-          cost = 0
-
           # P2 Thunes mekaniske verksted do not need to pay maintainance
           return 0 if owns_thunes_mekaniske?(route.train.owner)
 
-          route.all_hexes.each { |hex| cost += 10 if mountain?(hex) }
-          cost
+          route.all_hexes.count { |hex| mountain?(hex) } * 10
+        end
+
+        def revenue_for(route, stops)
+          track_types = route.chains.flat_map { |connections| connections[:paths] }.flat_map(&:track).uniq
+
+          raise GameError, 'Ships cannot run on land' if ship?(route.train) && track_types != [route.train.track_type]
+          raise GameError, 'Trains cannot run on water' if !ship?(route.train) && track_types.include?(:narrow)
+
+          cost = route_cost(route)
+          raise GameError, 'Cannot afford the fees for this route' if route.train.owner.cash < cost
```

Comment:
There's a method in the Game class called `check_other` which is where tests like this should be done.

Reply:
Updated